File: engine/static/wm/src/window_event.cpp
/*
 * This file is a part of Argus.
 * Copyright (c) 2019-2023, Max Roncace <mproncace@protonmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "argus/wm/window.hpp"
#include "argus/wm/window_event.hpp"
#include "internal/wm/pimpl/window.hpp"

namespace argus {
    WindowEvent::WindowEvent(WindowEventType subtype, Window &window, Vector2u resolution, Vector2i position,
            TimeDelta delta) :
            ArgusEvent{std::type_index(typeid(WindowEvent))},
            subtype(subtype),
            window(window),
            resolution(resolution),
            position(position),
            delta(delta) {
        window.pimpl->refcount.fetch_add(1);
    }

    WindowEvent::WindowEvent(WindowEventType subtype, Window &window) :
        WindowEvent(subtype, window, {}, {}, {}) {
    }

    WindowEvent::WindowEvent(const WindowEvent &rhs) :
        WindowEvent(rhs.subtype, rhs.window, rhs.resolution, rhs.position, rhs.delta) {
    }

    WindowEvent::~WindowEvent(void) {
        window.pimpl->refcount.fetch_sub(1);
    }
}

File: engine/auxiliary/wm_rustabi/src/argus/wm/window_event.rs
/*
 * This file is a part of Argus.
 * Copyright (c) 2019-2024, Max Roncace <mproncace@protonmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
use std::time::Duration;
use core_rustabi::argus::core::ArgusEvent;
use core_rustabi::core_cabi::argus_event_t;
use num_enum::{IntoPrimitive, UnsafeFromPrimitive};

use lowlevel_rustabi::argus::lowlevel::{Vector2i, Vector2u};

use crate::argus::wm::Window;
use crate::wm_cabi::*;

#[derive(Eq, Ord, PartialEq, PartialOrd, IntoPrimitive, UnsafeFromPrimitive)]
#[repr(u32)]
pub enum WindowEventType {
    Create = ARGUS_WINDOW_EVENT_TYPE_CREATE,
    Update = ARGUS_WINDOW_EVENT_TYPE_UPDATE,
    RequestClose = ARGUS_WINDOW_EVENT_TYPE_REQUEST_CLOSE,
    Minimize = ARGUS_WINDOW_EVENT_TYPE_MINIMIZE,
    Restore = ARGUS_WINDOW_EVENT_TYPE_RESTORE,
    Focus = ARGUS_WINDOW_EVENT_TYPE_FOCUS,
    Unfocus = ARGUS_WINDOW_EVENT_TYPE_UNFOCUS,
    Resize = ARGUS_WINDOW_EVENT_TYPE_RESIZE,
    Move = ARGUS_WINDOW_EVENT_TYPE_MOVE,
}

pub struct WindowEvent {
    handle: argus_window_event_t,
}

impl WindowEvent {
    pub fn get_subtype(&self) -> WindowEventType {
        unsafe {
            WindowEventType::unchecked_transmute_from(argus_window_event_get_subtype(self.handle))
        }
    }

    pub fn get_window(&self) -> Window {
        unsafe { Window::of(argus_window_event_get_window(self.handle)) }
    }

    pub fn get_resolution(&self) -> Vector2u {
        unsafe { argus_window_event_get_resolution(self.handle).into() }
    }

    pub fn get_position(&self) -> Vector2i {
        unsafe { argus_window_event_get_position(self.handle).into() }
    }

    pub fn get_delta(&self) -> Duration {
        unsafe { Duration::from_micros(argus_window_event_get_delta_us(self.handle)) }
    }
}

impl ArgusEvent for WindowEvent {
    fn get_type_id() -> &'static str {
        "window"
    }

    fn of(handle: argus_event_t) -> Self
    where Self: Sized
    {
        Self { handle }
    }

    fn get_handle(&self) -> argus_event_t {
        self.handle
    }
}
